src/resonate/actions.py

```diff
@@ -59,3 +68,180 @@ def with_options(
 @dataclass(frozen=True)
 class Sleep:
     seconds: int
+
+
+class Combinator(Generic[T]):
+    opts: Options = field(default=Options())
+
+    def __init__(self) -> None:
+        self.opts = Options(retry_policy=never())
+
+    def with_options(
+        self,
+        *,
+        durable: bool = True,
+        promise_id: str | None = None,
+        retry_policy: RetryPolicy | None = None,
+    ) -> Self:
+        """
+        Set options for the combinator.
+
+        Args:
+            durable (bool): Whether the promise is durable. Defaults to True.
+            promise_id (str | None): An optional identifier for the promise.
+            retry_policy (RetryPolicy | None): An optional retry policy for the promise.
+
+        Returns:
+            Self: The combinator instance with updated options.
+        """
+        self.opts = Options(
+            durable=durable,
+            promise_id=promise_id,
+            retry_policy=retry_policy if retry_policy is not None else never(),
+        )
+        return self
+
+    @abstractmethod
+    def done(self) -> bool:
+        pass
+
+    @abstractmethod
+    def result(self) -> Result[T, Exception]:
+        pass
+
+
+class All(Combinator[list[Any]]):
+    """
+    A combinator that waits for all promises to complete.
+
+    Attributes:
+        promises (list[Promise[Any]]): A list of promises to be combined.
+    """
+
+    def __init__(self, promises: list[Promise[Any]]) -> None:
+        super().__init__()
+        self.promises = promises
+
+    def done(self) -> bool:
+        """
+        Check if all promises are complete.
+
+        Returns:
+            bool: True if all promises are complete, False otherwise.
+        """
+        if len(self.promises) == 0:
+            return True
+
+        return all(p.done() for p in self.promises)
+
+    def result(self) -> Result[list[Any], Exception]:
```

Comment:
The typing here seems wrong. Isn't this supposed to be `list[Result[Any, Exception]]`? Same for `AllSettled`

Reply:
part of it is an artifact of trying to comply with the Combinators interface. the All combinator is right though, given that if one of the promises fail we want to return a single error. For all_settled I agree we could make it `list[Result[Any, Exception]]` once we make them private methods of the Scheduler